### PR TITLE
Polymorphism

### PR DESCRIPTION
**Pull request by lorchrob**
Support nodes with type parameters and node calls with type arguments.

### CODE REVIEW

File: src/lustre/typeCheckerContext.ml [hunk 1]
```diff
@@ -134,7 +141,7 @@ let rec lookup_ty_syn: tc_context -> LA.ident -> tc_type option
                   then Some ty
                   else lookup_ty_syn ctx uid
                | _ -> Some ty )
-  | None -> None
+  | None -> IMap.find_opt i (ctx.ty_syns)
```
Comment (daniel-larraz) [inline src/lustre/typeCheckerContext.ml:144]: Something is off. You are repeating the exact same search.

File: src/lustre/lustreAstHelpers.mli [hunk 2]
```diff
@@ -60,6 +60,9 @@ val apply_subst_in_type : (HString.t * expr) list -> lustre_type -> lustre_type
     to the expressions of (possibly dependent) type [t]
     AnyOp and Quantifier are not supported due to introduction of bound variables. *)
     
+val apply_type_subst_in_type : (HString.t * lustre_type) list -> lustre_type -> lustre_type
+(** [apply_subst_in_type s t] applies the substitution defined by association list [s]
```
Comment (daniel-larraz) [inline src/lustre/lustreAstHelpers.mli:64]:
```suggestion
(** [apply_type_subst_in_type s t] applies the substitution defined by association list [s]
```

File: doc/usr/source/2_input/1_lustre.rst [hunk 3]
```diff
@@ -1139,3 +1139,84 @@ was not included.
 
 Moreover, Kind 2 checks that any specified assumption in
 a ``any`` expression holds when model checking the component.
+
+Polymorphic nodes
+-----------------
+
+In some situations, the user may want to express multiple variations a node,
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply doc/usr/source/2_input/1_lustre.rst]: Addressed

File: src/lustre/lustreAstHelpers.mli (hunk 2 above)
Comment (lorchrob) [reply src/lustre/lustreAstHelpers.mli]: Addressed

File: src/lustre/typeCheckerContext.ml (hunk 1 above)
Comment (lorchrob) [reply src/lustre/typeCheckerContext.ml]: Addressed

**Pull request merged**